lm_scorer/models/gpt2.py

```diff
@@ -27,47 +43,41 @@ def _add_special_tokens(self, text: str) -> str:
     def _tokens_log_prob_for_batch(
         self, text: List[str]
     ) -> List[Tuple[torch.DoubleTensor, torch.LongTensor, List[str]]]:
-        device = self.model.device
-
-        outputs = []
-        for sentence in text:
-            # len(sentence) = seq_len + 2
-            sentence = self._add_special_tokens(sentence)
-            # len(tokens) = seq_len + 2
-            tokens: List[str] = self.tokenizer.tokenize(sentence)
-            # ids.shape = [1, seq_len + 2]
-            ids = torch.tensor(  # pylint: disable=not-callable
-                [self.tokenizer.convert_tokens_to_ids(tokens)],
-                device=device,
-                dtype=torch.long,
-            )
-
-            with torch.no_grad():
-                model_outputs: Tuple[torch.Tensor] = self.model(ids)
-
-            # pred_scores.shape = [1, seq_len + 2, vocab_size]
-            pred_scores = model_outputs[0].double()
-
-            # len(tokens) = seq_len + 1
-            tokens = tokens[1:]
-            # ids.shape = [1, seq_len + 1, vocab_size]
-            ids = ids[:, 1:]
-            # pred_scores.shape = [1, seq_len + 1, vocab_size]
-            pred_scores = pred_scores[:, :-1, :]
-
-            # ids_scores.shape = [1, seq_len + 1]
-            ids_scores = pred_scores.gather(2, ids.unsqueeze(2)).squeeze(2)
-            # log_prob.shape = [1, seq_len + 1]
-            log_probs = ids_scores - pred_scores.logsumexp(2)
-
-            log_probs = log_probs.squeeze(0)
-            ids = ids.squeeze(0)
-
-            log_probs = cast(torch.DoubleTensor, log_probs)
-            ids = cast(torch.LongTensor, ids)
-
-            output = (log_probs, ids, tokens)
+        outputs: List[Tuple[torch.DoubleTensor, torch.LongTensor, List[str]]] = []
+        if len(text) == 0:
+            return outputs
+
+        # TODO: Handle overflowing elements for long sentences
+        text = list(map(self._add_special_tokens, text))
+        encoding: BatchEncoding = self.tokenizer.batch_encode_plus(
+            text, return_tensors="pt",
+        )
+        with torch.no_grad():
+            ids = encoding["input_ids"].to(self.model.device)
+            nopad_mask = ids != self.tokenizer.pad_token_id
+            logits: torch.Tensor = self.model(ids)[0]
+
+        for sent_index in range(len(text)):
+            sent_nopad_mask = nopad_mask[sent_index]
+            # len(tokens) = len(text[sent_index]) + 1
+            first_pad_index = int(torch.sum(sent_nopad_mask).item())
+            sent_tokens = encoding.tokens(sent_index)[1:first_pad_index]
+
+            # sent_ids.shape = [len(text[sent_index]) + 1]
+            sent_ids = ids[sent_index, sent_nopad_mask][1:]
+            # logits.shape = [len(text[sent_index]) + 1, vocab_size]
+            sent_logits = logits[sent_index, sent_nopad_mask][:-1, self.nopad_vocab_idx]
```

Comment:
Isn't setting the logit for the padding token to `-inf` enough?
Something like `sent_logits.index_fill(1, self.tokenizer.pad_token_id, float("-inf"))` ?

Reply:
Yes, did not think of float("inf"). 
It is same as ?
`            sent_logits[:, self.tokenizer.pad_token_id] = float("-inf")
`

Reply:
Logically yes, one might be better optimized than the other tho.
Just run a little benchmark and see which one is faster.

Reply:
got this error 
`E           TypeError: index_fill_() received an invalid combination of arguments - got (int, int, float), but expected one of:
E            * (name dim, Tensor index, Tensor value)`
with 
`sent_logits.index_fill_(1, self.tokenizer.pad_token_id, float("-inf"))`
and I don't understand it.

Reply:
Does this work?
`sent_logits.index_fill_(1, pt.tensor([self.tokenizer.pad_token_id]), float("-inf"))`

It might be worth just using the code you suggested instead, it shouldn't really change anything.

Reply:
Yes it work, as you prefer for the syntax, I agree it won't change the perf.

Reply:
Let's use your suggestion then.